src/Perlang.Parser/NumberParser.cs

```diff
@@ -60,82 +60,121 @@ public static object Parse(NumericToken numericToken)
+                //
+                // Rationale: Two's complement: https://en.wikipedia.org/wiki/Two%27s_complement
+                if (negativeValue >= Int32.MinValue)
+                {
+                    return new ParsedNumber<int>((int)negativeValue);
+                }
+                else
+                {
+                    return new ParsedNumber<long>(negativeValue);
+                }
+            }
+            else if (parsedNumber.Value is long longValue)
             {
-                return (long)negativeValue;
+                return new ParsedNumber<long>(-longValue);
+            }
+            else if (parsedNumber.Value is ulong ulongValue)
+            {
+                // Again, this needs to be handled specially to ensure that numbers that fit in a `long` doesn't use
+                // BigInteger unnecessarily.
+                BigInteger negativeValue = -new BigInteger(ulongValue);
+
+                if (negativeValue >= Int64.MinValue)
+                {
+                    return new ParsedNumber<long>((long)negativeValue);
+                }
+                else
+                {
+                    // All negative numbers that are too big to fit in any of the smaller signed integer types will go
+                    // through this code path.
+                    return new ParsedNumber<BigInteger>(negativeValue);
+                }
             }
             else
             {
-                // All negative numbers that are too big to fit in any of the smaller signed integer types will go
-                // through this code path.
-                return negativeValue;
+                throw new ArgumentException($"Type {parsedNumber.Value.GetType().ToTypeKeyword()} not supported");
             }
         }
         else
         {
             throw new ArgumentException($"Type {value.GetType().ToTypeKeyword()} not supported");
         }
     }
+
+    internal class ParsedNumber<T> : IParsedNumber
+        where T : notnull
+    {
+        internal T Value { get; }
+
+        /// <inheritdoc cref="IParsedNumber.BitsUsed"/>
+        public long BitsUsed { get; }
+
+        object IParsedNumber.Value => Value;
+
+        public ParsedNumber(T value)
+        {
+            Value = value;
+
+            BitsUsed = value switch
+            {
+                double doubleValue => Math.ILogB(doubleValue),
+                int intValue => Math.ILogB(intValue),
+                uint uintValue => Math.ILogB(uintValue),
+                long longValue => Math.ILogB(longValue),
+                ulong ulongValue => Math.ILogB(ulongValue),
+                BigInteger bigintValue => bigintValue.GetBitLength(),
+                _ => throw new ArgumentException($"Unsupported numeric type encountered: {value.GetType().Name}")
+            };
+        }
+
+        public override string? ToString()
+        {
+            return Value.ToString();
+        }
+    }
```

Comment:
The changes in this file compose a significant refactoring. I think we'll extract this to a separate MR for reviewability, it's too entangled right now to be able to read this commit at a later stage and get _any_ idea of why things were done like this. :slightly_smiling_face: